### DISCUSSION TITLE
sqlalchemy.types has no attribute 'Uuid'

### DISCUSSION THREAD

**Discussion by dfinchdb** [sqlalchemy/sqlalchemy#10710] [Usage Questions]
Code run: 
from langchain.sql_database import SQLDatabase
db = SQLDatabase.from_databricks(catalog="hive_metastore", schema="foo", include_tables=["bar"]) 

Attribute Error: module 'sqlalchemy.types' has no attribute 'Uuid'
 
Environment producing error:
databricks-sql-connector==3.0.0
langchain==0.0.341
sqlalchemy==2.0.23

Working environment:
databricks-sql-connector==3.0.0
langchain==0.0.341
sqlalchemy==1.4.50 (also reported that worked with 2.0.12 but have not personally tested)

Stack trace:
AttributeError Traceback (most recent call last) File <command-1370453508995758>, line 2 1 from langchain.sql_database import SQLDatabase ----> 2 db = SQLDatabase.from_databricks(catalog="hive_metastore", schema="foo", include_tables=["bar"]) File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/langchain/utilities/sql_database.py:222, in SQLDatabase.from_databricks(cls, catalog, schema, host, api_token, warehouse_id, cluster_id, engine_args, **kwargs) 216 http_path = f"/sql/protocolv1/o/0/{cluster_id}" 218 uri = ( 219 f"databricks://token:{api_token}@{host}?" 220 f"http_path={http_path}&catalog={catalog}&schema={schema}" 221 ) --> 222 return cls.from_uri(database_uri=uri, engine_args=engine_args, **kwargs) File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/langchain/utilities/sql_database.py:125, in SQLDatabase.from_uri(cls, database_uri, engine_args, **kwargs) 123 """Construct a SQLAlchemy engine from URI.""" 124 _engine_args = engine_args or {} --> 125 return cls(create_engine(database_uri, **_engine_args), **kwargs) File <string>:2, in create_engine(url, **kwargs) File /databricks/python/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py:309, in deprecated_params.<locals>.decorate.<locals>.warned(fn, *args, **kwargs) 302 if m in kwargs: 303 _warn_with_version( 304 messages[m], 305 versions[m], 306 version_warnings[m], 307 stacklevel=3, 308 ) --> 309 return fn(*args, **kwargs) File /databricks/python/lib/python3.10/site-packages/sqlalchemy/engine/create.py:534, in create_engine(url, **kwargs) 530 u = _url.make_url(url) 532 u, plugins, kwargs = u._instantiate_plugins(kwargs) --> 534 entrypoint = u._get_entrypoint() 535 dialect_cls = entrypoint.get_dialect_cls(u) 537 if kwargs.pop("_coerce_config", False): File /databricks/python/lib/python3.10/site-packages/sqlalchemy/engine/url.py:655, in URL._get_entrypoint(self) 653 else: 654 name = self.drivername.replace("+", ".") --> 655 cls = registry.load(name) 656 # check for legacy dialects that 657 # would return a module with 'dialect' as the 658 # actual class 659 if ( 660 hasattr(cls, "dialect") 661 and isinstance(cls.dialect, type) 662 and issubclass(cls.dialect, Dialect) 663 ): File /databricks/python/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:341, in PluginLoader.load(self, name) 339 if [impl.name](http://impl.name/) == name: 340 self.impls[name] = impl.load --> 341 return impl.load() 343 raise exc.NoSuchModuleError( 344 "Can't load plugin: %s:%s" % (self.group, name) 345 ) File /databricks/python/lib/python3.10/site-packages/importlib_metadata/__init__.py:203, in EntryPoint.load(self) 198 """Load the entry point from its definition. If only a module 199 is indicated by the value, return that module. Otherwise, 200 return the named object. 201 """ 202 match = self.pattern.match(self.value) --> 203 module = import_module(match.group('module')) 204 attrs = filter(None, (match.group('attr') or '').split('.')) 205 return functools.reduce(getattr, attrs, module) File /usr/lib/python3.10/importlib/__init__.py:126, in import_module(name, package) 124 break 125 level += 1 --> 126 return _bootstrap._gcd_import(name[level:], package, level) File <frozen importlib._bootstrap>:1050, in _gcd_import(name, package, level) File <frozen importlib._bootstrap>:1027, in _find_and_load(name, import_) File <frozen importlib._bootstrap>:1006, in _find_and_load_unlocked(name, import_) File <frozen importlib._bootstrap>:688, in _load_unlocked(spec) File <frozen importlib._bootstrap_external>:883, in exec_module(self, module) File <frozen importlib._bootstrap>:241, in _call_with_frames_removed(f, *args, **kwds) File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/databricks/sqlalchemy/__init__.py:1 ----> 1 from databricks.sqlalchemy.base import DatabricksDialect 2 from databricks.sqlalchemy._types import TINYINT, TIMESTAMP, TIMESTAMP_NTZ 4 __all__ = ["TINYINT", "TIMESTAMP", "TIMESTAMP_NTZ"] File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/databricks/sqlalchemy/base.py:5 2 from typing import Any, List, Optional, Dict, Union, Collection, Iterable, Tuple 4 import databricks.sqlalchemy._ddl as dialect_ddl_impl ----> 5 import databricks.sqlalchemy._types as dialect_type_impl 6 from databricks import sql 7 from databricks.sqlalchemy._parse import ( 8 _describe_table_extended_result_to_dict_list, 9 _match_table_not_found_string, (...) 14 parse_column_info_from_tgetcolumnsresponse, 15 ) File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/databricks/sqlalchemy/_types.py:35 13 """This method is supposed to accept a Python type and return a string representation of that type. 14 But due to some weirdness in the way SQLAlchemy's literal rendering works, we have to return 15 the value itself because, by the time it reaches our custom type code, it's already been converted (...) 24 All fail without this hack in-place. I'm not sure why. But it works. 25 """ 26 return value 29 @compiles(sqlalchemy.types.Enum, "databricks") 30 @compiles(sqlalchemy.types.String, "databricks") 31 @compiles(sqlalchemy.types.Text, "databricks") 32 @compiles(sqlalchemy.types.Time, "databricks") 33 @compiles(sqlalchemy.types.Unicode, "databricks") 34 @compiles(sqlalchemy.types.UnicodeText, "databricks") ---> 35 @compiles(sqlalchemy.types.Uuid, "databricks") 36 def compile_string_databricks(type_, compiler, **kw): 37 """ 38 We override the default compilation for Enum(), String(), Text(), and Time() because SQLAlchemy 39 defaults to incompatible / abnormal compiled names (...) 49 But all of these types will be compiled to STRING in Databricks SQL 50 """ 51 return "STRING" AttributeError: module 'sqlalchemy.types' has no attribute 'Uuid'

**Comment by gordthompson** [comment]
Also reported here

https://github.com/databricks/databricks-sql-python/issues/291

```python
from sqlalchemy.types import Uuid
```

works in SQLAlchemy 2.0 but not in SQLAlchemy 1.4. The dialect must have been updated for 2.0.
